lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1527,6 +1527,60 @@ static void diagRecursivePropertyAccess(const Expr *E, const DeclContext *DC) {
   const_cast<Expr *>(E)->walk(walker);
 }
 
+/// Whether or not this closure captures self weakly
+static bool closureHasWeakSelfCapture(const AbstractClosureExpr *ACE) {
+  if (auto closureExpr = dyn_cast<ClosureExpr>(ACE)) {
+    if (auto selfDecl = closureExpr->getCapturedSelfDecl()) {
+      return selfDecl->getType()->is<WeakStorageType>();
+    }
+  }
+
+  return false;
+}
+
+/// Whether or not implicit self DREs in this closure are incorrect,
+/// and need to be looked-up manually. This applies to weak self
+/// closures in Swift 5 mode, where the AST is incorrect.
+static bool
+closureHasIncorrectImplicitSelfDecls(const AbstractClosureExpr *ACE) {
+  if (ACE->getASTContext().LangOpts.isSwiftVersionAtLeast(6)) {
+    return false;
+  }
+
+  return closureHasWeakSelfCapture(ACE);
+}
+
+// Returns true if this is an implicit self expr
+static bool isImplicitSelf(const Expr *E) {
+  auto *DRE = dyn_cast<DeclRefExpr>(E);
+
+  if (!DRE || !DRE->isImplicit())
+    return false;
+
+  ASTContext &Ctx = DRE->getDecl()->getASTContext();
+  return DRE->getDecl()->getName().isSimpleName(Ctx.Id_self);
+}
+
+// When inside a closure with incorrect implicit self decls
+// (any closure with a weak self capture, in Swift 5 mode),
+// we have to manually lookup what self refers to at the location
+// of a given DRE instead of using the decl in the AST (which is incorrect).
+static ValueDecl *valueDeclForDRE(const DeclRefExpr *DRE,
```

Comment:
Same here - we could call this `getCapturedSelfDecl` or similar to make it clear that we're not somehow switching out what `self` refers to in the closure, we're just doing lookup for the captured `self` (for the purpose of the diagnostics pass)?